src/ellipse.rs

```diff
@@ -0,0 +1,261 @@
+    /// A new ellipse from center, radii, and x_rotation.
+    ///
+    /// The returned ellipse will be the result of taking a circle, stretching it by the `radii`
+    /// along the x and y axes, then rotating it clockwise by `x_rotation` radians, before finally
+    /// translating the center to `center`.
+    #[inline]
+    pub fn new(center: impl Into<Point>, radii: impl Into<Vec2>, x_rotation: f64) -> Ellipse {
+        let Point { x: cx, y: cy } = center.into();
+        let Vec2 { x: rx, y: ry } = radii.into();
+        Ellipse::private_new(Vec2 { x: cx, y: cy }, rx, ry, x_rotation)
+    }
+
+    /// This gives us an internal method without any type conversions.
+    #[inline]
+    fn private_new(center: Vec2, scale_x: f64, scale_y: f64, x_rotation: f64) -> Ellipse {
+        Ellipse {
+            inner: Affine::translate(center)
+                * Affine::rotate(x_rotation)
+                * Affine::scale_non_uniform(scale_x, scale_y),
+        }
+    }
+
+    /// Create an ellipse from an affine transformation of the unit circle.
+    #[inline]
+    pub fn from_affine(affine: Affine) -> Self {
+        Ellipse { inner: affine }
+    }
+
+    // Getters and setters.
+
+    /// Returns the center of this ellipse.
+    #[inline]
+    pub fn center(&self) -> Point {
+        let Vec2 { x: cx, y: cy } = self.inner.get_translation();
+        Point { x: cx, y: cy }
+    }
+
+    /// Sets the center of this ellipse.
+    #[inline]
+    pub fn with_center(self, new_center: Point) -> Ellipse {
+        let Point { x: cx, y: cy } = new_center;
+        Ellipse {
+            inner: self.inner.set_translation(Vec2 { x: cx, y: cy }),
+        }
+    }
+
+    /// Returns the two radii of this ellipse.
+    ///
+    /// The first number is the horizontal radius and the second is the vertical radius, before
+    /// rotating by `x_rotation`.
+    pub fn radii(&self) -> Vec2 {
+        self.inner.svd().0
+    }
+
+    /// Set the radii of this ellipse.
+    #[must_use]
+    pub fn with_radii(self, new_radii: Vec2) -> Ellipse {
+        let rotation = self.inner.svd().1;
+        let translation = self.inner.get_translation();
+        Ellipse::private_new(translation, new_radii.x, new_radii.y, rotation)
+    }
+
+    /// The amount (in radians) that the ellipse should be rotated by (clockwise).
+    ///
+    /// This allows all possible ellipses to be drawn by always starting with an ellipse with the
+    /// two radii on the x and y axes.
+    pub fn x_rotation(&self) -> f64 {
+        self.inner.svd().1
+    }
+
+    /// Set the amount (in radians) that the ellipse should be rotated by (clockwise).
+    pub fn with_x_rotation(self, new_x_rotation: f64) -> Ellipse {
+        let scale = self.inner.svd().0;
+        let translation = self.inner.get_translation();
+        Ellipse::private_new(translation, scale.x, scale.y, new_x_rotation)
+    }
+}
+
+// We need to implement this ourselves as multiple affine maps represent the same ellipse. If `==`
```

Comment:
It seems to me this will generate `false` in almost all nontrivial cases, because of floating point rounding error. We have been agonizing for a while about "fuzzy equality" but haven't adopted a solution. I'm ok with this as it is but am wondering whether it might be simpler to just use == on the affine.

Reply:
@derekdreery I think this comment and the one below need resolution and then we're good to go. :)

Reply:
You're right, and even then there are multiple representations of the ellipse (e.g. changing the sign of the radii is the identity). I will just use the derived impl.